### PR TITLE
HTML API: Add method to report depth of currently-matched element.

### PR DESCRIPTION
**Pull request by dmsnell**
Trac ticket: Core-61255.

The HTML Processor maintains a stack of open elements, where every element, every `#text` node, every HTML comment, and other node is pushed and popped while traversing the document. The "depth" of each of these nodes represents how deep that stack is where the node appears. Unfortunately this information isn't exposed to calling code, which has led different projects to attempt to calculate this value externally. This isn't always trivial, but the HTML Processor could make it so by exposing the internal knowledge in a new method.

In this patch the `get_current_depth()` method returns just that. Since the processor always exists within a context, the depth includes nesting from the always-present `html` element and also the `body`, since currently the HTML Processor only supports parsing in the IN BODY context.

This means that the depth reported for the `DIV` in `<div>` is 3, not 1, because its breadcrumbs path is `HTML > BODY > DIV`.

### CODE REVIEW

**Review by sirreal — APPROVED** [review]
This looks good.

File: src/wp-includes/html-api/class-wp-html-processor.php [hunk 1]
```diff
@@ -1263,6 +1263,17 @@ public function get_tag() {
 		}
 	}
 
+	/**
+	 * Return how deep the currently-matched element is in the HTML document.
+	 *
+	 * @since 6.6.0
+	 *
+	 * @return int The depth of the currently-matched element.
```
Comment (sirreal) [inline src/wp-includes/html-api/class-wp-html-processor.php]: Should we be talking about "element" here? It doesn't need to be an element we're at… should this talk about "token" or more generally "position of the processor in the document"?

File: tests/phpunit/tests/html-api/wpHtmlProcessor.php [hunk 2]
```diff
@@ -348,8 +348,15 @@ public static function data_unsupported_special_in_body_tags() {
 	public function test_reports_proper_element_depth_in_body( $html_with_target_element, $depth_at_element ) {
 		$processor = WP_HTML_Processor::create_fragment( $html_with_target_element );
 
-		$this->assertTrue(
-			$processor->next_tag( array( 'class_name' => 'target' ) ),
+		while ( $processor->next_tag() ) {
+			if ( $processor->has_class( 'target' ) ) {
+				break;
```
Comment (dmsnell) [inline tests/phpunit/tests/html-api/wpHtmlProcessor.php]: why this change?

File: src/wp-includes/html-api/class-wp-html-processor.php (hunk 1 above)
Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-processor.php]: I think we should say "node". These can be `#text` nodes here. Is "position" right? I think in places we've used "location in the document" but also in the HTML Processor I feel like we're free to talk more specifically about nodes and elements since that's what it operates on, vs. text tokens.

Comment (sirreal) [reply src/wp-includes/html-api/class-wp-html-processor.php]: Node sounds good.

File: tests/phpunit/tests/html-api/wpHtmlProcessor.php (hunk 2 above)
Comment (sirreal) [reply tests/phpunit/tests/html-api/wpHtmlProcessor.php]: I thought the other form was broken, causing the test to fail. That's the only reason. I may be mistaken, feel free to revert.

File: src/wp-includes/html-api/class-wp-html-processor.php (hunk 1 above)
Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-processor.php]: my own confusing is clearer now: I had un-pushed work, including the method itself. I've sent this out in d54d71fc94

apparently I had already chosen "of the current location" in my work, so I'm leaving it at that unless we prefer "node"

File: tests/phpunit/tests/html-api/wpHtmlProcessor.php (hunk 2 above)
Comment (dmsnell) [reply tests/phpunit/tests/html-api/wpHtmlProcessor.php]: oops, this is a known bug I forgot about. we claim to support `class_name` but don't

**Comment by dmsnell** [comment]
Merged in [[58191]](https://core.trac.wordpress.org/changeset/58191)
https://github.com/wordpress/wordpress-develop/commit/5d52f19635fe0989fc2e46d447e6c9c220a204e0
